### DISCUSSION TITLE
OpenFGA tree-sitter grammar

### DISCUSSION THREAD

**Discussion by matoous** [openfga/community#288] [Show and tell]
Hi 👋 

I was missing highlights for the `fga` files in my IDE so after some toying around, here's a tree-sitter grammar for OpenFGA authorization schema files in case somebody finds it useful: https://github.com/matoous/tree-sitter-fga.

If the OpenFGA org members were interested I would be happy to move it under `github.com/openfga`.

<img width="860" alt="highlights" src="https://github.com/openfga/community/assets/15747583/e2faea1e-eb4f-42b4-869e-7c2a47f84716">

**Comment by rhamzeh** [comment]
Hi @matoous! Thanks for sharing this!

You can also find the antlr grammar for parsing OpenFGA DSL in [language](https://github.com/openfga/language), we also offer language definition in [Prism.js](https://github.com/openfga/frontend-utils/blob/main/src/tools/prism/language-definition.ts), [Monaco](https://github.com/openfga/frontend-utils/blob/main/src/tools/monaco/language-definition.ts) and [VS Code](https://github.com/openfga/vscode-ext/blob/main/syntaxes/openfga.tmLanguage.json).

The actual colors for the [OpenFGA Dark theme can be found here](https://github.com/openfga/frontend-utils/blob/main/src/theme/supported-themes/openfga-dark.ts).

Please feel free to re-use any parts that make parsing easier on your end. Also do share your project on our CNCF Slack channel, found [here](https://openfga.dev/docs/community#slack-cncf-community)

**Reply by matoous** [reply]
Thanks @rhamzeh! ~Looking at the DSL I can see that the `condition` highlights are subpar, will try to adjust that.~ Updated, there are probably still some mismatches but the DSL also seems to be quite permissive so now the highlights should work better.

**Comment by aaguiarz** [comment]
@matoous what IDE are you using?

**Reply by matoous** [reply]
[helix](https://github.com/helix-editor/helix), so tree-sitter was the only option.

**Reply by aaguiarz** [reply]
Cool. What's the easiest way to install it in helix to try it? Are you aware of other IDEs that have built-in support for tree-sitter?

**Reply by matoous** [reply]
You will need highlight queries which can be done by adding:
```
(call_expression
  function: (identifier) @function)

(call_expression
  function: (selector_expression
    field: (identifier) @function.method))

((type_identifier) @type.builtin
  (#match? @type.builtin "^(string|int|map|uint|list|timestamp|bool|duration|double|ipaddress)$"))

(condition_declaration
  name: (identifier) @function)

(version) @number

[
  "*"
  "/"
  "%"
  ">>"
  "<<"
  "&"
  "&^"
] @operator

[
  "+"
  "-"
  "|"
  "^"
] @operator

[
  "or"
  "and"
  "but not"
] @operator

[
  "model"
  "schema"
  "type"
  "relations"
  "define"
  "from"
] @keyword

[
  "condition"
] @keyword.function

(comment) @comment
```
to `~/.config/helix/runtime/queries/fga/highlights.scm`. (~plenty of room for improvement there~ updated, probably can still be improved but should suffice for now). And you will need to configure FGA language by adding:
```
[[language]]
name = "fga"
scope = "source.fga"
file-types = ["fga"]
comment-token = "#"
indent = { tab-width = 2, unit = "  " }
injection-regex = "fga"

[[grammar]]
name = "fga"
source = { git = "https://github.com/matoous/tree-sitter-fga", rev = "be5b2ba518e42efe6ff9354053e7a34326b144ba" }
```
to `~/.config/helix/languages.toml`. That should be it.

`nvim` has support for tree-sitter, frequently configured via `nvim-treesitter`: https://github.com/nvim-treesitter/nvim-treesitter.
[lapce](https://github.com/lapce/lapce) seems to be using tree-sitter. VSCode has currently active discussion about switching to tree-sitter as well: https://github.com/microsoft/vscode/issues/50140. Last that I know of is [zed](https://zed.dev/).

**Reply by aaguiarz** [reply]
@matoous would it make sense to add that documentation of how to configure it for Helix on your repository's documentation? Over time other members could add docs explaining how to enable it in other editors.

I'm adding the project to the Community Projects list [here](https://github.com/openfga/community?tab=readme-ov-file#community-projects).

Thanks a lot!

**Reply by matoous** [reply]
I would be also happy to open up a PR against helix so it's there out of the box, same for nvim. First I wanted to make sure this doesn't collide with other plans and there's actually demand for it.

Otherwise yes, happy to update the docs. Thanks for suggestion!

**Comment by aaguiarz** [comment]
We have in our roadmap add support for other IDEs, next was Jetbrain's editors. We might tackle `tree-sitter`-based ones later and we can collaborate on getting it to the openfga organization. 

We are kind of conservative of adding new projects to the openfga org until we can make sure there's enough interest and that the OpenFGA maintainers can actually maintain them in the long term.
